File: src/main/java/me/sargunvohra/leveluphp/command/RootCommand.java
package me.sargunvohra.leveluphp.command;

import static net.minecraft.command.Commands.literal;

import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import lombok.experimental.UtilityClass;
import lombok.val;
import me.sargunvohra.leveluphp.LevelUpHp;
import me.sargunvohra.leveluphp.capability.PlayerLevelHandler;
import net.minecraft.command.CommandSource;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.TextComponentString;

@UtilityClass
public class RootCommand {

  private final int MIN_PERMISSION = 2;

  public LiteralArgumentBuilder<CommandSource> register() {
    val base = literal(LevelUpHp.MOD_ID);
    base.then(SetIntCommand.register("setxp", MIN_PERMISSION, PlayerLevelHandler::setXp));
    base.then(SetIntCommand.register("addxp", MIN_PERMISSION, PlayerLevelHandler::addXp));
    base.then(SetIntCommand.register("setlevel", MIN_PERMISSION, PlayerLevelHandler::setLevel));
    base.then(SetIntCommand.register("addlevel", MIN_PERMISSION, PlayerLevelHandler::addLevel));
    base.then(GetCommand.register("check", MIN_PERMISSION, RootCommand::checkCommand));
    return base;
  }

  private ITextComponent checkCommand(PlayerLevelHandler handler) {
    return new TextComponentString(
        String.format(
            "level: %d, xp: %d/%d", handler.getLevel(), handler.getXp(), handler.xpTarget()));
  }
}

File: src/main/java/me/sargunvohra/leveluphp/command/RootCommand.kt
package me.sargunvohra.leveluphp.command

import com.mojang.brigadier.builder.LiteralArgumentBuilder
import me.sargunvohra.leveluphp.LevelUpHp
import me.sargunvohra.leveluphp.core.PlayerLevelHandler
import net.minecraft.command.CommandSource
import net.minecraft.command.Commands.literal
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.TextComponentString

class RootCommand {
    companion object {

        private const val MIN_PERMISSION = 2

        fun register(): LiteralArgumentBuilder<CommandSource> {
            val base = literal(LevelUpHp.MOD_ID)
            listOf(
                SetIntCommand.register("setxp", MIN_PERMISSION) { obj, xp -> obj.xp = xp },
                SetIntCommand.register("addxp", MIN_PERMISSION) { obj, xp -> obj.xp += xp },
                SetIntCommand.register("setlevel", MIN_PERMISSION) { obj, level -> obj.level = level },
                SetIntCommand.register("addlevel", MIN_PERMISSION) { obj, levels -> obj.level += levels },
                GetCommand.register("check", MIN_PERMISSION) { checkCommand(it) }
            ).forEach { base.then(it) }
            return base
        }

        private fun checkCommand(handler: PlayerLevelHandler): ITextComponent {
            return TextComponentString(
                String.format("level: %d, xp: %d/%d", handler.level, handler.xp, handler.xpTarget)
            )
        }
    }
}
